### DISCUSSION TITLE
psycopg dialect (psycopg3) is lacking Numeric and Float types

### DISCUSSION THREAD

**Discussion by devkral** [sqlalchemy/sqlalchemy#11637] [Usage Questions]
I contributed to the databasez an abstraction around sqlalchemy (mostly for transactions when they didn't exist in sqlalchemy) and noticed
that the psycopg postgresql dialect is missing the sqltypes:

Numeric and Float

Is there a reason behind this? I think it is a bug.

Workaround used by the abstraction:

https://github.com/dymmond/databasez/blob/cea07fe93183d47d283a925b40588fea9ac62210/databasez/backends/dialects/psycopg.py

**Comment by CaselIT** [comment]
Hi,

I'm not sure I follow. What you mean by missing?
This table works perfectly fine in psycopg
```py
import sqlalchemy as sa

e = sa.create_engine('postgresql+psycopg://...')

t = sa.Table(
    "foo",
    sa.MetaData(),
    sa.Column("a", sa.Float),
    sa.Column("b", sa.Numeric),
)

t.metadata.create_all(e)
with e.connect() as conn:
    conn.execute(t.insert(), [{"a": 1.0, "b": 3.0}])
    conn.commit()

    print(conn.execute(t.select()).all())
t.metadata.drop_all(e)
```

**Reply by devkral** [reply]
Thank you, that is a valuable information. Then there is an automatic fallback in sqlalchemy to the default dialect which the abstraction doesn't have and has to provide.

But still a question to the sqlalchemy devs. Is it really intentional that nothing like:

``` python
class _PGBigInteger(sqltypes.BigInteger):
    render_bind_cast = True

```
is used for Float or Numeric?

File: [lib/sqlalchemy/dialects/postgresql/psycopg.py](https://github.com/sqlalchemy/sqlalchemy/blob/800932af467109f06c0196c42ae86272a5d7f96a/lib/sqlalchemy/dialects/postgresql/psycopg.py)

**Reply by CaselIT** [reply]
I guess it wasn't deemed necessary. Looking as asyncpg it seems that it defines them, so it could make sense to defined custom subtypes, mainly to support the `render_bind_cast`?

Have you encountered issues caused by the missing casts?

**Reply by devkral** [reply]
yes, it makes it harder to use the dialect direct for formatting sql.
The same issue seems to be in psycopg2 too.

**Reply by CaselIT** [reply]
I don't understand the use case. What do you mean by formatting sql?
Can you provide an example of what you are trying to do and why you require something like https://github.com/dymmond/databasez/blob/cea07fe93183d47d283a925b40588fea9ac62210/databasez/backends/dialects/psycopg.py ?
